### ISSUE TITLE
Tile generation line visual bug

### ISSUE THREAD

**Issue by JoyYou19** [Droggelbecher/bevy-fast-tilemap#21]
So I am using this plugin to create a pretty big tilemap and I seem to have run into a very interesting visual problem. I believe it might be due to how the chunks are loaded for tiles but if I move around sometimes I can subtly see black lines go accross my screen. At first I thought that it was no big deal, but then I was walking around and randomly stopped on a spot where those black lines were covering the screen and they were staying there the whole time.

Maybe you can look into it, but in the meantime I will just ignore it.
[
![image_2024-03-13_154842477](https://github.com/Droggelbecher/bevy-fast-tilemap/assets/47606406/7ff3cdde-1554-4dd2-82c0-6b3919d4ef28)
](url)

**Comment by Droggelbecher** [comment]
Hm, perhaps a rounding issue in the tile position calculation, what is your tile size and zoom factor?

**Comment by Droggelbecher** [comment]
Also, does it happen for the whole map or only on one side? If the latter, its likely related to floating point rounding

**Comment by JoyYou19** [comment]
I believe it happens to the whole map, and I am quite lost at where the problem could be. For some reason I thought maybe that is some sort of screen tearing or something related to the render mode, but that is not the case. Also I tried to test it without the perlin noise and I still am able to see those lines.

The tile size is 80x80 pixels, and I haven't zoomed in at all. Those are the default camera settings. And in this case the texture is 4x the size than the pixels that are visible. I have scaled it from 20x20 per tile to 80x80 per tile. 

```rust
use bevy::prelude::*;

use bevy::math::{uvec2, vec2, vec3};
use bevy_fast_tilemap::{FastTileMapPlugin, Map, MapBundleManaged};

use crate::rendering::unload_chunks;
use crate::AppState;
use noise::{NoiseFn, Perlin};

pub struct WorldGenPlugin;

pub const WORLDSIZE: UVec2 = uvec2(1280, 1280);

impl Plugin for WorldGenPlugin {
    fn build(&self, app: &mut App) {
        app.add_plugins(FastTileMapPlugin::default())
            .add_systems(OnExit(AppState::Finished), generate_world)
            .add_systems(FixedUpdate, unload_chunks.run_if(in_state(AppState::Game)));
    }
}

#[derive(Resource, Default)]
pub struct WorldGen {
    pub size: UVec2,
    pub tile_data: Vec<Vec<u32>>,
    pub world_grid: Vec<Vec<u32>>,
}

fn generate_world(
    mut commands: Commands,
    asset_server: Res<AssetServer>,
    mut materials: ResMut<Assets<Map>>,
) {
    // Create map with (10 * 128) ^ 2 tiles or 1,638,400 tiles.
    let perlin = Perlin::new(1);
    // Create a 2D array to store tile indices
    let mut tile_struct = WorldGen {
        size: WORLDSIZE,
        tile_data: vec![vec![0; WORLDSIZE.x as usize]; WORLDSIZE.y as usize],
        ..default()
    };
    let tile_indices = &mut tile_struct.tile_data;

    let mut map = Map::builder(
        // Map size (tiles)
        uvec2(tile_struct.size.y, tile_struct.size.x),
        // Tile atlas
        asset_server.load("tilemaps/spr_grass.png"),
        // Tile size (pixels)
        vec2(80., 80.),
    )
    .build_and_initialize(|m| {
        for y in 0..m.size().y {
            for x in 0..m.size().x {
                // Calculate noise value for the current position
                let noise_value = perlin.get([x as f64 / 5.0, y as f64 / 5.0]); // Adjust frequency as needed

                // Map noise value to tile index
                let tile_index = if noise_value < 0.0 { 3 } else { 4 };

                // Store the tile index in the 2D array
                tile_indices[y as usize][x as usize] = tile_index;

                // Set tile index for the current position
                m.set(x, y, tile_index);
            }
        }
    });

    let mut m = map.indexer_mut();
    for y in 0..m.size().y {
        for x in 0..m.size().x {
            let mut tile_index = 3;
            if tile_indices[y as usize][x as usize] != 3 {
                tile_index = tile_bitmasking_floor(x as u32, y as u32, &tile_indices);
            }

            m.set(x, y, tile_index);
        }
    }

    let mut bundle = MapBundleManaged::new(map, materials.as_mut());
    // Higher z value means "closer to the camera"
    bundle.transform = Transform::default().with_translation(vec3(0., 0., -100.));

    commands.spawn(bundle);

    commands.insert_resource(tile_struct);
}

fn tile_bitmasking_floor(x: u32, y: u32, tile_indices: &Vec<Vec<u32>>) -> u32 {
    let width = tile_indices[0].len() as u32;
    let height = tile_indices.len() as u32;

    // Calculate the offsets for the surrounding tiles
    let north_tile = if y > 0 {
        if tile_indices[(y - 1) as usize][x as usize] > 3 {
            1
        } else {
            0
        }
    } else {
        0
    };
    let south_tile = if y < height - 1 {
        if tile_indices[(y + 1) as usize][x as usize] > 3 {
            1
        } else {
            0
        }
    } else {
        0
    };
    let west_tile = if x > 0 {
        if tile_indices[y as usize][(x - 1) as usize] > 3 {
            1
        } else {
            0
        }
    } else {
        0
    };
    let east_tile = if x < width - 1 {
        if tile_indices[y as usize][(x + 1) as usize] > 3 {
            1
        } else {
            0
        }
    } else {
        0
    };

    // Perform the bitmasking calculation
    let index = north_tile + 2 * west_tile + 4 * east_tile + 8 * south_tile;

    // Check for index 0 and return accordingly
    if index == 0 {
        return 3;
    } else {
        return index + 4;
    }
}
```

**Comment by Droggelbecher** [comment]
Hm I have trouble reproducing this, maybe its something about the tilemap itself (perhaps the upscaling?). I'm currently trying to reproduce here: https://github.com/Droggelbecher/bevy-fast-tilemap/blob/fix/visual-line-bug/examples/test_line_bug.rs

Also, which OS are you using?

**Comment by JoyYou19** [comment]
I am on Windows, and I can include the texture that I am using here. Also these tiles are offset by a bunch of empty ones since that way it is easier to calculate the bitmasking for me.
![spr_grass](https://github.com/Droggelbecher/bevy-fast-tilemap/assets/47606406/51d7deb6-2e10-4ce8-861a-598f0e8bc884)

**Comment by Droggelbecher** [comment]
Thanks! Seems I can reproduce it, but it depends on the precise camera position, I think this might indeed be a floating point thing but needs more experimenting, I'll keep you posted :)

**Comment by Droggelbecher** [comment]
Ok here are some insights: There were a couple of FP related issues going on:

The first one is that the inverse of your tilesize (1/80.0) is not precisely representable in float32:
```
In [2]: np.float32(1/80.0)*80.0
Out[2]: 1.0000000149011612
```
This means that tiles end up shifted by some small amount from their correct position which means for high coordinate values and the right conditions, two neighbouring tiles may get rounded to slightly mismatching positions, leaving the observed visual gap.
This issue luckily can be easily worked around by choosing a power of two as tile size such as 64.0 or 128.0.
I had the feeling this reduces the issues for me at least for "small" maps like 1280x1280, but alas doesn't fully resolve it.

Another issue is that 1/1280 (the number of tiles in one direction) has the exact same problem. Again, easy to work around: Use a power of two for your map size. 

Finally, there wore some small inaccuracies in how bordering pixels were handled that I'm addressing in https://github.com/Droggelbecher/bevy-fast-tilemap/tree/fix/visual-line-bug

TL;DR: Could you try:
* Making both your map size and tile size be a power of two (eg 2048x2048 x 128x128)? That should already help a lot
* Check out the branch I mentioned above, that should further reduce the errors.

I'm not 100% sure yet this resolves all of the visual glitches but at least for me they seem to become pretty rare with this.

**Comment by JoyYou19** [comment]
Okay, so I tested this with a bunch of tilesets and I can safely say that in most cases your new branch on github fixes most of the issues. I also experimented with powers of 2 and it seems that it is fixed. But now I am at a point where I don't necessarily know how to include your fixed version in my project. I am fairly new to Bevy and the rust ecosystem as a whole, so I am quite lost here. Thank you for your input and time!

**Comment by Droggelbecher** [comment]
No worries, I found some more changes, once I'm done, Ill make a new release, with some luck still this week.

**Comment by Droggelbecher** [comment]
No worries, I found some more changes I want to do, once I'm done, I'll make a new release, with some luck still this week :)

**Comment by Droggelbecher** [comment]
New version 0.7.1 is released with a fundamentally different approach to the calculations that should hopefully lead to much less lines between the tiles you saw, probably also for non-power-of-two tile sizes. Hope this helps :)
